dsp/modules/ollama.py

```diff
@@ -0,0 +1,168 @@
+from dsp.modules.lm import LM
+from typing import Any, Literal, Optional
+
+import os, multiprocessing, datetime, hashlib
+import requests
+
+import json
+
+def post_request_metadata(model_name, prompt):
+    """Creates a serialized request object for the Ollama API."""
+    timestamp = datetime.datetime.now().timestamp()
+    id_string = str(timestamp) + model_name + prompt
+    hashlib.sha1().update(id_string.encode("utf-8"))
+    id_hash = hashlib.sha1().hexdigest()
+    return {
+        "id": f"chatcmpl-{id_hash}",
+        "object": "chat.completion",
+        "created": int(timestamp),
+        "model": model_name
+    }
+
+
+class OllamaLocal(LM):
+    """Wrapper around a locally hosted Ollama model (API: https://github.com/jmorganca/ollama/blob/main/docs/modelfile.md#valid-parameters-and-values and https://github.com/jmorganca/ollama/blob/main/docs/api.md#generate-a-completion). 
+    Returns dictionary info in the OpenAI API style (https://platform.openai.com/docs/api-reference/chat/object).
+
+    Args:
+        model (str, optional): Name of Ollama model. Defaults to "llama2".
+        model_type (Literal["chat", "text"], optional): The type of model that was specified. Mainly to decide the optimal prompting strategy. Defaults to "text".
+        **kwargs: Additional arguments to pass to the API.
+    """
+    
+    def __init__(
+        self,
+        model: str = "llama2", 
+        model_type: Literal["chat", "text"] = None,
+        **kwargs,
+    ):
+        super().__init__(model)
+        self.provider = "ollama"
+
+        self.base_url = "http://localhost:11434"  # where the model is hosted
```

Comment:
This is hardcoded but I merged for now as it's probably going to be easy to refactor a bit